index/scorch/segment/zap/mem.go

```diff
@@ -0,0 +1,99 @@
+//  Copyright (c) 2017 Couchbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 		http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package zap
+
+import (
+	"io"
+	"io/ioutil"
+	"sync/atomic"
+
+	mmap "github.com/edsrzf/mmap-go"
+)
+
+var (
+	mmapCurrentBytes int64
+
+	// Ignore mmap failures and fallback to regular file access.
+	MmapIgnoreErrors bool
+
+	// Optional, maximum number of bytes to mmap before fallback to regular file access.
+	MmapMaxBytes int64
+)
+
+type zapStats struct {
+	// Total number of bytes mapped into memory with mmap.
+	MmapCurrentBytes int64
+}
+
+type mmapOwner interface {
+	readMM(int64, int64) []byte
+}
+
+// Stats returns memory usage details for open zap segments in this process.
+func Stats() zapStats {
+	return zapStats{
+		MmapCurrentBytes: atomic.LoadInt64(&mmapCurrentBytes),
+	}
+}
+
+func (s *SegmentBase) readMem(x, y uint64) []byte {
+	if s.mem == nil {
+		return s.mmapOwner.readMM(int64(x), int64(y))
+	}
+	return s.mem[x:y]
+}
+
+func (s *Segment) readMM(x, y int64) []byte {
+	if s.mm == nil {
+		data, _ := ioutil.ReadAll(io.NewSectionReader(s.f, x, y-x))
```

Comment:
ignoring the err here is preferred?

I may not have understood the thing here- hence a doubt,

Isn't this ioutil.ReadAll()  conveniently overrides the MmapMaxBytes or any such memory limits set? Or you were worried only about the mmap memory usage spikes?

My understanding is that - s.mm will be nil if your file size is bigger than MmapMaxBytes, and then
what if the size of the file is big, and then ioutil.ReadAll reads & copies the whole content to memory? And lets assume there are concurrent queries and mutations are happening in back ground?

Won't this approach will also have high memory requirements due to these "data" slices holding the whole file contents unlike that in mmap?

Reply:
> Or you were worried only about the mmap memory usage spikes?

Correct.

So the underlying issue is not so much memory usage, but address space limitations. If a 32-bit process can only access 2GB of memory addresses, then mmap will start failing after 2GB of zap segments are mapped. And other methods like pthread_create also start failing because they need to reserve address space for stack, etc.

By limiting the address space used by mmap (via `MmapMaxBytes`), we ensure that part of the address space is available to the golang runtime. The runtime itself will reserve space for its own heap, and those addresses will be managed by the golang GC.

> s.mm will be nil if your file size is bigger than MmapMaxBytes

Yep, so if mmap fails or is over the limit, then `s.mm` is nil and that signals to the code to fallback to regular file access.

> what if the size of the file is big, and then ioutil.ReadAll reads & copies the whole content to memory

From what I can tell, `readMem` will never try to read the entire file. It only tries to read certain slices which are currently required. These will be copied onto the golang heap, but then as soon as they're not being used anymore the GC will reap them and re-use that memory for reads and segments.

> And lets assume there are concurrent queries and mutations are happening in back ground?

AFAICT, zap segments are immutable and zap only opens them in read-only mode.

>Won't this approach will also have high memory requirements due to these "data" slices holding the whole file contents unlike that in mmap?

Yes, basically we're trading off address space reservations in exchange for golang heap usage. By leaving part of the address space free, we allow golang heap to expand and the GC to manage which parts of the file are held in memory (vs assuming the entire thing can be mapped in like it does on 64-bit OS).

Reply:
Great questions @sreekanth-cb  and answers @tmm1 this helps my understanding already.

Reply:
> ignoring the err here is preferred?

Not sure about this one. It seems like if the file was able to be opened, and we checked its stat/size already, then reads should all succeed. I'm open to logging errors in this path, but I'm not sure where to send them.